src/Microsoft.AspNetCore.Blazor.Razor.Extensions/BlazorRuntimeNodeWriter.cs

```diff
@@ -522,57 +400,42 @@ public override void WriteComponentAttribute(CodeRenderingContext context, Compo
             {
                 throw new InvalidOperationException("Unexpected node type " + node.Children[0].GetType().FullName);
             }
-            
+
             context.CodeWriter.Write(");");
             context.CodeWriter.WriteLine();
         }
 
         public override void WriteReferenceCapture(CodeRenderingContext context, RefExtensionNode node)
         {
-            if (node.IsComponentCapture)
-            {
-                // AddComponentReferenceCapture calls can be inserted directly because they are
-                // already in the desired place in the IR
-                WriteAddReferenceCaptureCall(context, node);
-            }
-            else
-            {
-                // AddElementReferenceCapture calls must appear in the RenderTree as children of the element
-                // being captured, so store the RefCapture until we're ready to write its children.
-                // We could instead modify RefLoweringPass to insert the RefExtensionNode at the right
-                // place, but this is harder because it would involve understanding which HtmlContent
-                // node represents the end of the tag. It's simpler to do it here because here we are
-                // already doing HTML parsing to understand the structure.
-                _currentElementRefCaptures.Add(node);
-            }
-        }
+            var codeWriter = context.CodeWriter;
 
-        private SourceSpan? CalculateSourcePosition(
-            SourceSpan? razorTokenPosition,
-            TextPosition htmlNodePosition)
-        {
-            if (razorTokenPosition.HasValue)
-            {
-                var razorPos = razorTokenPosition.Value;
-                return new SourceSpan(
-                    razorPos.FilePath,
-                    razorPos.AbsoluteIndex + htmlNodePosition.Position,
-                    razorPos.LineIndex + htmlNodePosition.Line - 1,
-                    htmlNodePosition.Line == 1
-                        ? razorPos.CharacterIndex + htmlNodePosition.Column - 1
-                        : htmlNodePosition.Column - 1,
-                    length: 1);
-            }
-            else
+            var methodName = node.IsComponentCapture
+                ? nameof(BlazorApi.RenderTreeBuilder.AddComponentReferenceCapture)
+                : nameof(BlazorApi.RenderTreeBuilder.AddElementReferenceCapture);
+            codeWriter
+                .WriteStartMethodInvocation($"{_scopeStack.BuilderVarName}.{methodName}")
+                .Write((_sourceSequence++).ToString())
+                .WriteParameterSeparator();
+
+            const string refCaptureParamName = "__value";
```

Comment:
The new approach that I took for the HTML element and component nodes just builds the correct ordering into the design. This way features like attributes and ref can't be broken by putting the nodes in the wrong order.